src/vs/base/common/event.ts

```diff
@@ -116,6 +116,11 @@ export namespace Event {
 			}
 		});
 
+		const onDidDispose = originalEvent[_onDidDispose];
+		if (typeof onDidDispose === 'function') {
+			once(onDidDispose)(() => emitter.dispose());
+		}
+
```

Comment:
Before, `snapshot` was effectively side-effect free. Now, it subscribes to `onDidDispose` without ever unsubscribing. Isn't this a leak?

I think this should only be done on `onFirstListenerAdd` (and the returned disposable should be cleaned on `onLastListenerRemove`).

Reply:
> Before, snapshot was effectively side-effect free.

I believe the side-effect of `snapshot` is a listener on the original event which cannot be unhooked (when having bad downlevel listener). 

> Now, it subscribes to onDidDispose without ever unsubscribing. Isn't this a leak?

It uses `once` to auto-unsubscribe when the original event is being disposed. I don't believe that it can be done with first/last. This is about the `emitter` that's created when you call `snapshot` and that should be cleared independent of its event having listeners or not.